CONTRIBUTING.md

```diff
@@ -62,6 +62,24 @@ When a cluster is first created with a cluster config file, there is no master n
 * The machine controller creates resources (Machines etc)
 * Pivot the apiserver and the machine controller in to the cluster.
 
+#### Configurable Machine Setup
+
+While not mandatory, it is suggested for new providers to support configurable machine setups for creating new machines. 
+This is to allow flexibility in what startup scripts are used and what versions are supported instead of hardcoding startup scripts into the machine controller.
+A yaml file (e.g. `machine_setup_configs.yaml`) should hold the list of valid machine setup configs, and the yaml file will be volume mounted into the machine controller using a ConfigMap named `machine-setup`. 
+
+A config type should define a set of parameters that can be taken from the machine object being created, and map those parameters to startup scripts and other relevant information. 
+For example in GCE, the OS, machine roles, and version info are the parameters that map to a GCP image path and metadata (which contains the startup script).
+
+When creating a new machine, there should be a check for whether the machine setup is supported. 
+This is done by looking through the valid configs parsed out of the yaml for a config with matching parameters.
+If a match is found, then the machine can be created with the startup script found in the config.
```

Comment:
I think it would be helpful to explicitly call out that the implied logic that if it is not found, it is not a supported machine configuration.